### ISSUE TITLE
Attention dropout stays active at eval/inference in the Graphormer encoder (`dropout_p` passed unconditionally to `F.scaled_dot_product_attention`)

### ISSUE THREAD

**Issue by ansonkejin-design** [coleygroup/ms-pred#38]
## Summary

`src/ms_pred/graphormer/multihead_attention.py` (current `main`, `forward`, ~line 70) passes the attention dropout probability to `F.scaled_dot_product_attention` unconditionally:

```python
attn = F.scaled_dot_product_attention(
    q, k, v,
    attn_mask=attn_bias,
    dropout_p=self.dropout,   # always the constructor value, regardless of module mode
    is_causal=False,
)
```

`F.scaled_dot_product_attention` is a functional API with **no `training` flag** — whenever `dropout_p > 0` it applies dropout, and it is the caller's responsibility to zero it outside training. `MultiheadAttention.forward` never checks `self.training`, so **attention dropout stays active after `model.eval()`**.

## Consequences

1. **Non-reproducible inference.** Every forward pass advances the global torch RNG (once per Graphormer layer; the dropout mask is generated inside the fused SDPA kernel, so it is easy to miss). Output spectra therefore depend on the process RNG state: the same input produces different spectra depending on what ran before it in the same process (batch composition, input order, unrelated requests). Reruns are bitwise identical only when the execution plan is identical.
   - Related: `--seed` in `glacier/predict_smis_joint.py` (~line 78) is parsed but never wired to `torch.manual_seed` / numpy / CUDA seeds, so runs are not seed-controlled either.
2. **Noisier eval than intended.** Benchmark/eval spectra are computed with p=0.1 attention-dropout noise on top of the deterministic model output. Dataset-level mean metrics average this out, so it is invisible in aggregate tables, but every per-spectrum output is perturbed.

## Minimal reproduction

```python
import torch
from ms_pred.glacier import dataset as glacier_dataset, joint_model, predict_smis_joint as psj
from torch.utils.data import DataLoader

model = joint_model.JointModel.load_from_checkpoint(<any GLACIER joint ckpt>)
model.eval(); model.freeze()

tp = glacier_dataset.TreeProcessor(pe_embed_k=model.pe_embed_k, root_encode="graphormer",
                                   embed_elem_group=model.embed_elem_group,
                                   multi_hop_max_dist=model.multi_hop_max_dist)
entries = [(smi, name, ce, adduct, "Orbitrap", "")]
ds = psj.GraphormerFeatDataset(entries, tree_processor=tp, sort_for_batching=False, feat_cache_size=0)
batch = next(iter(DataLoader(ds, num_workers=0, collate_fn=psj.collate_predict, batch_size=1)))

torch.manual_seed(1); out1 = model.predict_inten_frag_batch_sparse(batch, 100)
torch.manual_seed(2); out2 = model.predict_inten_frag_batch_sparse(batch, 100)
# out1 != out2 (differences far beyond float epsilon)
# also: torch.get_rng_state() advances during every eval forward
```

Workaround while this is open — zero the attribute after loading:

```python
from ms_pred.graphormer.multihead_attention import MultiheadAttention
for m in model.modules():
    if isinstance(m, MultiheadAttention):
        m.dropout = 0.0
```

## Measured impact (torch 2.3.0+cu118, CUDA GPU, 50 molecules × 2 adducts × 3 collision energies = 300 requests, sparse_k=100)

- Same requests, before vs after zeroing eval attention dropout: **0/300 spectra identical**; per-spectrum binned-cosine (0.01 Da bins) **median 0.995, P5 0.895, min 0.57** — i.e. eval-time dropout perturbs every predicted spectrum.
- With the fix: outputs become seed-independent and bitwise identical across processes; per-request results become independent of batch composition/order at batch_size=1 (determinism was the motivation for measuring this in the first place — we hit it while building a content-addressed prediction cache for a downstream application).

## Suggested one-line fix

```python
dropout_p=self.dropout if self.training else 0.0,
```

Training behavior is unchanged (`self.training` is True during training); eval stops consuming RNG and matches standard PyTorch module semantics (`nn.Dropout` no-ops in eval mode).

Happy to send a PR if useful — thanks for the great work on this repo.

**Comment by RuiXiWangTW** [comment]
Thanks for pointing it out! We merged the fix into the main branch.
